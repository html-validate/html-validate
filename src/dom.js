'use strict';

const DOMNode = require('./domnode').default;

class DOM {
	constructor(){
		this.root = DOMNode.rootNode();
		this.active = this.root;
	}

	pushActive(node){
		this.active = node;
	}

	popActive(){
		this.active = this.active.parent;
	}

	getActive(){
		return this.active;
	}
}

module.exports = DOM;
