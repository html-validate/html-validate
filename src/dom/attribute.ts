import { Location } from "../context";
import { DynamicValue } from "./dynamic-value";

interface StaticAttribute {
	readonly value: string | null;
}

interface DynamicAttribute {
	readonly value: DynamicValue;
}

/**
 * Type guard for [[Attribute]] testing if the value is static (missing value is
 * considered static as it is a fixed known value).
 *
 * @public
 */
export function isStaticAttribute(attr: Attribute | null): attr is Attribute & StaticAttribute {
	return Boolean(attr && attr.isStatic);
}

/**
 * Type guard for [[Attribute]] testing if the value is dynamic.
 *
 * @public
 */
export function isDynamicAttribute(attr: Attribute | null): attr is Attribute & DynamicAttribute {
	return Boolean(attr && attr.isDynamic);
}

/**
 * DOM Attribute.
 *
 * Represents a HTML attribute. Can contain either a fixed static value or a
 * placeholder for dynamic values (e.g. interpolated).
 */
export class Attribute {
	/** Attribute name */
	public readonly key: string;
	public readonly value: string | DynamicValue | null;
	public readonly keyLocation: Location;
	public readonly valueLocation: Location | null;
	public readonly originalAttribute?: string;

	/**
	 * @param key - Attribute name.
	 * @param value - Attribute value. Set to `null` for boolean attributes.
	 * @param keyLocation - Source location of attribute name.
	 * @param valueLocation - Source location of attribute value.
	 * @param originalAttribute - If this attribute was dynamically added via a
	 * transformation (e.g. vuejs `:id` generating the `id` attribute) this
	 * parameter should be set to the attribute name of the source attribute (`:id`).
	 */
	public constructor(
		key: string,
		value: null | string | DynamicValue | null,
		keyLocation: Location,
		valueLocation: Location | null,
		originalAttribute?: string
	) {
		this.key = key;
		this.value = value;
		this.keyLocation = keyLocation;
		this.valueLocation = valueLocation;
		this.originalAttribute = originalAttribute;

		/* force undefined to null */
		if (typeof this.value === "undefined") {
			this.value = null;
		}
	}

	/**
	 * Flag set to true if the attribute value is static.
	 */
	public get isStatic(): boolean {
		return !this.isDynamic;
	}

	/**
	 * Flag set to true if the attribute value is dynamic.
	 */
	public get isDynamic(): boolean {
		return this.value instanceof DynamicValue;
	}

	/**
	 * Test attribute value.
	 *
	 * @param pattern - Pattern to match value against. RegExp or a string (===)
	 * @param dynamicMatches - If true `DynamicValue` will always match, if false
	 * it never matches.
	 * @returns `true` if attribute value matches pattern.
	 */
	public valueMatches(pattern: RegExp, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: string, dynamicMatches?: boolean): boolean;
	public valueMatches(pattern: RegExp | string, dynamicMatches: boolean = true): boolean {
		if (this.value === null) {
			return false;
		}

		/* dynamic values matches everything */
		if (this.value instanceof DynamicValue) {
			return dynamicMatches;
		}

		/* test value against pattern */
		if (pattern instanceof RegExp) {
			return this.value.match(pattern) !== null;
		} else {
			return this.value === pattern;
		}
	}
}
