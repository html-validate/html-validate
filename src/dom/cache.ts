/**
 * @public
 */
/* eslint-disable-next-line @typescript-eslint/no-empty-object-type -- will be augmented by other modules */
export interface DOMNodeCache {}
