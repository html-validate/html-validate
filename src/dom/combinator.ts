export enum Combinator {
	DESCENDANT,
	CHILD,
	ADJACENT_SIBLING,
	GENERAL_SIBLING,
}

export function parseCombinator(combinator: string): Combinator {
	switch (combinator){
	case undefined:
	case null:
	case "":
		return Combinator.DESCENDANT;
	case ">":
		return Combinator.CHILD;
	case "+":
		return Combinator.ADJACENT_SIBLING;
	case "~":
		return Combinator.GENERAL_SIBLING;
	default:
		throw new Error(`Unknown combinator "${combinator}"`);
	}
}
