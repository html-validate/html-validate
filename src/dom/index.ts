export { Attribute } from "./attribute";
export { HtmlElement, NodeClosed } from "./htmlelement";
export { DOMNode } from "./domnode";
export { DOMTokenList } from "./domtokenlist";
export { DOMTree } from "./domtree";
export { DynamicValue } from "./dynamic-value";
export { NodeType } from "./nodetype";
export { TextNode } from "./text";
