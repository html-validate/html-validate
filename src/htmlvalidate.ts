import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
