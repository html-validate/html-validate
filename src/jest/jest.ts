import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
