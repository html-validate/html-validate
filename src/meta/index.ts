export { MetaTable } from "./table";
export { MetaElement, MetaData, PropertyExpression } from "./element";
export { Validator } from "./validator";
