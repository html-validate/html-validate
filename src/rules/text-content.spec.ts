import { DynamicValue, HtmlElement } from "../dom";
import { HtmlValidate } from "../htmlvalidate";
import "../jest";
import { TextContent } from "../meta";

function processElement(node: HtmlElement): void {
	if (node.hasAttribute("bind-text")) {
		node.appendText(new DynamicValue(""), {
			filename: "mock",
			line: 1,
			column: 1,
			offset: 0,
			size: 1,
		});
	}
}

describe("rule text-content", () => {
	let htmlvalidate: HtmlValidate;

	beforeAll(() => {
		htmlvalidate = new HtmlValidate({
			elements: [
				{
					"text-unset": {},
					"text-none": {
						textContent: TextContent.NONE,
					},
					"text-default": {
						textContent: TextContent.DEFAULT,
					},
					"text-required": {
						textContent: TextContent.REQUIRED,
					},
					"text-accessible": {
						textContent: TextContent.ACCESSIBLE,
					},
					input: {
						void: true,
						textContent: TextContent.ACCESSIBLE,
					},
				},
			],
			rules: { "text-content": "error" },
		});
	});

	describe("text unset", () => {
		it("should not report error when meta is missing", () => {
			expect.assertions(1);
			const markup = [
				"<text-missing></text-missing>",
				"<text-missing>foobar</text-missing>",
				'<text-missing><span aria-label="foobar"></span></text-missing>',
			].join("");
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error when textContent is unset", () => {
			expect.assertions(1);
			const markup = [
				"<text-unset></text-unset>",
				"<text-unset>foobar</text-unset>",
				'<text-unset><span aria-label="foobar"></span></text-unset>',
			].join("");
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});
	});

	describe("text none", () => {
		it("should not report when element is missing text", () => {
			expect.assertions(1);
			const markup = "<text-none></text-none>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report when element have only interelement whitespace", () => {
			expect.assertions(1);
			const markup = "<text-none>\n</text-none>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should report error when element have static text", () => {
			expect.assertions(2);
			const markup = "<text-none>foobar</text-none>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-none> must not have text content");
		});

		it("should report error when element have dynamic text", () => {
			expect.assertions(2);
			const report = htmlvalidate.validateString('<text-none bind-text="dynamic"></text-none>', {
				processElement,
			});
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-none> must not have text content");
		});

		it("should report error when child element have static text", () => {
			expect.assertions(2);
			const markup = "<text-none><span>foobar</span></text-none>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-none> must not have text content");
		});

		it("should report error when child element have dynamic text", () => {
			expect.assertions(2);
			const markup = '<text-none><span bind-text="dynamic"></span></text-none>';
			const report = htmlvalidate.validateString(markup, {
				processElement,
			});
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-none> must not have text content");
		});
	});

	describe("text default", () => {
		it("should not report when element is missing text", () => {
			expect.assertions(1);
			const markup = "<text-default></text-default>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report when element have only interelement whitespace", () => {
			expect.assertions(1);
			const markup = "<text-default>\n</text-default>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error when element have static text", () => {
			expect.assertions(1);
			const markup = "<text-default>foobar</text-default>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error when element have dynamic text", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString(
				'<text-default bind-text="dynamic"></text-default>',
				{ processElement }
			);
			expect(report).toBeValid();
		});

		it("should not report error when child element have text", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString(
				"<text-default><span>foobar</span></text-default>"
			);
			expect(report).toBeValid();
		});
	});

	describe("text required", () => {
		it("should report when element is missing text", () => {
			expect.assertions(2);
			const markup = "<text-required></text-required>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-required> must have text content");
		});

		it("should report when element have only interelement whitespace", () => {
			expect.assertions(2);
			const markup = "<text-required>\n</text-required>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-required> must have text content");
		});

		it("should not report error when element have static text", () => {
			expect.assertions(1);
			const markup = "<text-required>foobar</text-required>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error when element have dynamic text", () => {
			expect.assertions(1);
			const markup = '<text-required bind-text="dynamic"></text-required>';
			const report = htmlvalidate.validateString(markup, { processElement });
			expect(report).toBeValid();
		});

		it("should not report error when child element have static text", () => {
			expect.assertions(1);
			const markup = "<text-required><span>foobar</span></text-required>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error when child element have dynamic text", () => {
			expect.assertions(1);
			const markup = '<text-required><span bind-text="dynamic">foobar</span></text-required>';
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});
	});

	describe("text accessible", () => {
		it("should report when element is missing text", () => {
			expect.assertions(2);
			const markup = "<text-accessible></text-accessible>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-accessible> must have accessible text");
		});

		it("should report when element have only interelement whitespace", () => {
			expect.assertions(2);
			const markup = "<text-accessible>\n</text-accessible>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-accessible> must have accessible text");
		});

		it("should report when children with text is aria-hidden", () => {
			expect.assertions(2);
			const markup = '<text-accessible><span aria-hidden="true">foobar</span></text-accessible>';
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toHaveError("text-content", "<text-accessible> must have accessible text");
		});

		describe("should report error when element has", () => {
			it.each`
				description                            | markup
				${"empty aria-label"}                  | ${'<text-accessible aria-label=""></text-accessible>'}
				${"empty aria-label (child)"}          | ${'<text-accessible><span aria-label=""></span></text-accessible>'}
				${"default text overwritten to empty"} | ${'<input type="submit" value="">'}
			`("$description", ({ markup }) => {
				expect.assertions(1);
				const report = htmlvalidate.validateString(markup);
				expect(report).toBeInvalid();
			});
		});

		it("should not report error when element have text", () => {
			expect.assertions(1);
			const markup = "<text-accessible>foobar</text-accessible>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error when child element have text", () => {
			expect.assertions(1);
			const markup = "<text-accessible><span>foobar</span></text-accessible>";
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error when element is aria-hidden", () => {
			expect.assertions(1);
			const markup = '<text-accessible aria-hidden="true"></text-accessible>';
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error parent is aria-hidden", () => {
			expect.assertions(1);
			const markup = '<div aria-hidden="true"><text-accessible></text-accessible></div>';
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error when element have dynamic text", () => {
			expect.assertions(1);
			const markup = '<text-accessible bind-text="dynamic"></text-accessible>';
			const report = htmlvalidate.validateString(markup, { processElement });
			expect(report).toBeValid();
		});

		describe("should not report error when element has", () => {
			it.each`
				description                               | markup
				${"aria-label"}                           | ${'<text-accessible aria-label="foobar"></text-accessible>'}
				${"aria-label (child)"}                   | ${'<text-accessible><span aria-label="foobar"></span></text-accessible>'}
				${"aria-labelledby"}                      | ${'<text-accessible aria-labelledby="reference"></text-accessible>'}
				${"aria-labelledby (child)"}              | ${'<text-accessible><span aria-labelledby="reference"></span></text-accessible>'}
				${"img with alt text"}                    | ${'<text-accessible><img alt="foobar"></img></text-accessible>'}
				${'default text (<input type="submit">)'} | ${'<input type="submit">'}
				${'default text (<input type="reset">)'}  | ${'<input type="reset">'}
			`("$description", ({ markup }) => {
				expect.assertions(1);
				const report = htmlvalidate.validateString(markup);
				expect(report).toBeValid();
			});
		});
	});

	describe("should contain documentation", () => {
		it.each(Object.values(TextContent))("%s", async (textContent: string) => {
			expect.assertions(1);
			const context = {
				tagName: "my-element",
				textContent,
			};
			const docs = await htmlvalidate.getContextualDocumentation({
				ruleId: "text-content",
				context,
			});
			expect(docs).toMatchSnapshot();
		});
	});
});
