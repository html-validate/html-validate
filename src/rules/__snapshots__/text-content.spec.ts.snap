// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule text-content should contain contextual documentation accessible 1`] = `
{
  "description": "The \`<my-element>\` element must have accessible text.",
  "url": "https://html-validate.org/rules/text-content.html",
}
`;

exports[`rule text-content should contain contextual documentation default 1`] = `
{
  "description": "The textual content for this element is not valid.",
  "url": "https://html-validate.org/rules/text-content.html",
}
`;

exports[`rule text-content should contain contextual documentation none 1`] = `
{
  "description": "The \`<my-element>\` element must not have textual content.",
  "url": "https://html-validate.org/rules/text-content.html",
}
`;

exports[`rule text-content should contain contextual documentation required 1`] = `
{
  "description": "The \`<my-element>\` element must have textual content.",
  "url": "https://html-validate.org/rules/text-content.html",
}
`;

exports[`rule text-content should contain documentation 1`] = `
{
  "description": "The textual content for this element is not valid.",
  "url": "https://html-validate.org/rules/text-content.html",
}
`;
